### PR TITLE
test: validate MCOPY opcode for FIP-0094

### PR DESCRIPTION
**Pull request by BigLep**
This PR adds an itest to validate that the MCOPY work of FIP-0094 is having its desired effect.

## Related Issues
https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0094.md
https://github.com/filecoin-project/FIPs/pull/1047
https://github.com/filecoin-project/builtin-actors/pull/1572

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title conforms with [contribution conventions](https://github.com/filecoin-project/lotus/blob/master/CONTRIBUTING.md#pr-title-conventions)
- [ ] Update CHANGELOG.md or signal that this change does not need it per [contribution conventions](https://github.com/filecoin-project/lotus/blob/master/CONTRIBUTING.md#changelog-management)
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by BigLep** [comment]
To be clear, this isn't work I have done.  I am only creating the PR so that we have something tangible on the nv24 track project board.

**Comment by BigLep** [comment]
Per https://filecoinproject.slack.com/archives/CP50PPW2X/p1728030176551119?thread_ts=1728008390.565939&cid=CP50PPW2X @snissn has requested some debug help from @rvagg .

File: itests/fevm_test.go [hunk 1]
```diff
@@ -1299,3 +1303,46 @@ func TestEthGetTransactionCount(t *testing.T) {
...
+	// Also check for the specific error message
+	expectedErrMsg := "contract reverted (33)"
+	require.Contains(t, err.Error(), expectedErrMsg)
+
+	// wait for the upgrade
+	client.WaitTillChain(ctx, kit.HeightAtLeast(nv24epoch+5))
+
+	// should be able to deploy and call the contract now
+	fromAddr, contractAddr = client.EVM().DeployContractFromFilename(ctx, filenameActor)
+	_, _, err = client.EVM().InvokeContractByFuncName(ctx, fromAddr, contractAddr, "optimizedCopy(string)", []byte("testdata"))
```
Comment (Stebalien) [inline itests/fevm_test.go:1346]: This isn't the correct format for the parameters, unless I'm mistaken. This needs to be encoded as a variable-length string. See https://docs.soliditylang.org/en/latest/abi-spec.html#use-of-dynamic-types.

In this case, we need (concatenated):

1. 32 byte word indicating the offset of `testdata` (in bytes). So, `int256(32).to_bigendian()`.
2. The length of the data `int256(len("testdata")).to_bigendian()`.
3. "testdata", padded with zeros on the right to the next multiple of 32 bytes.

I usually use tools like: https://abi.hashex.org/

That gives us

```
683f3ac5000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000087465737464617461000000000000000000000000000000000000000000000000
```

Which includes the function selector at the beginning, you'll want to remove that.

**Comment by BigLep** [comment]
For visibility and cross-linking, there is an mcopy test in builtin-actors as well: https://github.com/filecoin-project/builtin-actors/pull/1583

File: itests/fevm_test.go (hunk 1 above)
Comment (snissn) [reply itests/fevm_test.go]: thank you for explaining this multiple times!

File: itests/fevm_test.go [hunk 2]
```diff
@@ -1310,6 +1315,59 @@ func TestEthGetTransactionCount(t *testing.T) {
...
+	// Also check for the specific error message
+	expectedErrMsg := "undefined instruction (35)"
+	require.Contains(t, err.Error(), expectedErrMsg)
+
+	// wait for the upgrade
+	client.WaitTillChain(ctx, kit.HeightAtLeast(nv24epoch+5))
+
+	// should be able to deploy and call the contract now
+	fromAddr, contractAddr = client.EVM().DeployContractFromFilename(ctx, filenameActor)
+	_, _, err = client.EVM().InvokeContractByFuncName(ctx, fromAddr, contractAddr, "optimizedCopy(bytes)", inputArgument)
```
Comment (Stebalien) [inline itests/fevm_test.go:1367]: Can we make sure we got the correct output value?

File: itests/fevm_test.go [hunk 3]
```diff
@@ -1310,6 +1315,59 @@ func TestEthGetTransactionCount(t *testing.T) {
...
+	)
+
+	ctx, cancel, client := kit.SetupFEVMTest(t, upgradeSchedule)
+	defer cancel()
+
+	// try to deploy the contract before the upgrade, expect an error somewhere' in deploy or in call,
+	// if the error is in deploy we may need to implement DeployContractFromFilename here where we can
+	// assert an error
+
+	// 0000000000000000000000000000000000000000000000000000000000000020: The offset for the bytes argument (32 bytes).
```
Comment (rvagg) [inline itests/fevm_test.go:1341]: nice, thanks for including these details, very helpful to have this in docs!

File: itests/fevm_test.go [hunk 4]
```diff
@@ -1310,6 +1315,59 @@ func TestEthGetTransactionCount(t *testing.T) {
...
+
+	filenameActor := "contracts/mcopy/MCOPYTest.hex"
+	fromAddr, contractAddr := client.EVM().DeployContractFromFilename(ctx, filenameActor)
+	_, _, err = client.EVM().InvokeContractByFuncName(ctx, fromAddr, contractAddr, "optimizedCopy(bytes)", inputArgument)
+	// We expect an error here due to the contract reverting or another issue.
+	require.Error(t, err)
+
+	// Also check for the specific error message
+	expectedErrMsg := "undefined instruction (35)"
+	require.Contains(t, err.Error(), expectedErrMsg)
```
Comment (rvagg) [inline itests/fevm_test.go:1360]:
```suggestion
	// We expect an error here due to MCOPY not being available in this network version
	require.ErrorContains(t, err, "undefined instruction (35)")
```

File: itests/fevm_test.go [hunk 5]
```diff
@@ -1310,6 +1315,59 @@ func TestEthGetTransactionCount(t *testing.T) {
...
+
+	// wait for the upgrade
+	client.WaitTillChain(ctx, kit.HeightAtLeast(nv24epoch+5))
+
+	// should be able to deploy and call the contract now
+	fromAddr, contractAddr = client.EVM().DeployContractFromFilename(ctx, filenameActor)
+	_, _, err = client.EVM().InvokeContractByFuncName(ctx, fromAddr, contractAddr, "optimizedCopy(bytes)", inputArgument)
+	require.NoError(t, err)
```
Comment (rvagg) [inline itests/fevm_test.go:1368]:
```suggestion
	result, _, err := client.EVM().InvokeContractByFuncName(ctx, fromAddr, contractAddr, "optimizedCopy(bytes)", inputArgument)
	require.NoError(t, err)
	require.Equal(t, inputArgument, result)
```

File: itests/fevm_test.go [hunk 6]
```diff
@@ -1310,6 +1315,59 @@ func TestEthGetTransactionCount(t *testing.T) {
...
+	// 0000000000000000000000000000000000000000000000000000000000000020: The offset for the bytes argument (32 bytes).
+	// 0000000000000000000000000000000000000000000000000000000000000008: The length of the bytes data (8 bytes for "testdata").
+	// 7465737464617461000000000000000000000000000000000000000000000000: The hexadecimal representation of "testdata", padded to 32 bytes.
+	hexString := "000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000087465737464617461000000000000000000000000000000000000000000000000"
+
+	// Decode the hex string into a byte slice
+	inputArgument, err := hex.DecodeString(hexString)
+	if err != nil {
+		log.Fatalf("Failed to decode hex string: %v", err)
+	}
```
Comment (rvagg) [inline itests/fevm_test.go:1350]:
```suggestion
	require.NoError(t, err)
```

File: itests/fevm_test.go (hunk 2 above)
Comment (snissn) [reply itests/fevm_test.go]: thanks! @rvagg added that

File: itests/fevm_test.go [hunk 7]
```diff
@@ -8,21 +8,26 @@ import (
 	"encoding/hex"
 	"encoding/json"
 	"fmt"
+	"log"
```
Comment (rvagg) [inline itests/fevm_test.go:11]:
```suggestion
```

**Review by rvagg — APPROVED** [review]

**Comment by rvagg** [comment]
Awesome, thanks @snissn, all done

**Pull request merged**